gdnative-core/src/class.rs

```diff
@@ -60,71 +69,149 @@ pub trait NativeClassMethods: NativeClass {
+            (gd_api.godot_method_bind_ptrcall)(set_library, native_script, args.as_mut_ptr(), std::ptr::null_mut());
+
+            let owner = T::Base::construct();
 
-            Self {
-                this: self.this,
-                _marker: PhantomData,
+            assert_ne!(std::ptr::null_mut(), owner.to_sys(), "base object should not be null");
+
+            let mut args: [*const libc::c_void; 1] = [native_script as *const _];
+            (gd_api.godot_method_bind_ptrcall)(object_set_script, owner.to_sys(), args.as_mut_ptr(), std::ptr::null_mut());
+
+            let script_ptr = (gd_api.godot_nativescript_get_userdata)(owner.to_sys()) as *const libc::c_void;
+
+            assert_ne!(std::ptr::null(), script_ptr, "script instance should not be null");
+
+            let script = T::UserData::clone_from_user_data_unchecked(script_ptr);
+
+            object::unref(native_script);
+
+            Instance {
+                owner,
+                script,
             }
         }
     }
 
-    fn get_impl(&self) -> &RefCell<T> {
-        unsafe {
-            let api = get_api();
-            let ud = (api.godot_nativescript_get_userdata)(self.this);
-            &*(ud as *const _ as *const RefCell<T>)
-        }
+    pub fn into_base(self) -> T::Base {
+        self.owner
     }
 
-    #[doc(hidden)]
-    pub unsafe fn sys(&self) -> *mut sys::godot_object {
-        self.this
+    /// Calls a function with a NativeClass instance and its owner, and returns its return
+    /// value. Can be used on reference counted types for multiple times.
+    pub fn map<F, U>(&self, op: F) -> Result<U, <T::UserData as Map>::Err>
+    where
+        T::Base: Clone,
+        T::UserData: Map,
+        F: FnOnce(&T, T::Base) -> U,
+    {
+        self.script.map(|script| op(script, self.owner.clone()))
+    }
+
+    /// Calls a function with a NativeClass instance and its owner, and returns its return
+    /// value. Can be used on reference counted types for multiple times.
+    pub fn map_mut<F, U>(&self, op: F) -> Result<U, <T::UserData as MapMut>::Err>
+    where
+        T::Base: Clone,
+        T::UserData: MapMut,
+        F: FnOnce(&mut T, T::Base) -> U,
+    {
+        self.script.map_mut(|script| op(script, self.owner.clone()))
+    }
+
+    /// Calls a function with a NativeClass instance and its owner, and returns its return
+    /// value. Can be used for multiple times via aliasing. For reference-counted types
+    /// behaves like the safe `map`, which should be preferred.
+    pub unsafe fn map_aliased<F, U>(&self, op: F) -> Result<U, <T::UserData as Map>::Err>
+    where
+        T::UserData: Map,
+        F: FnOnce(&T, T::Base) -> U,
+    {
+        self.script.map(|script| op(script, T::Base::from_sys(self.owner.to_sys())))
+    }
+
+    /// Calls a function with a NativeClass instance and its owner, and returns its return
+    /// value. Can be used for multiple times via aliasing. For reference-counted types
```

Comment:
```suggestion
    /// value. Can be used for multiple items via aliasing. For reference-counted types
```

Reply:
By "for multiple times" I meant you could do this:

```rust
unsafe {
    instance.map_aliased(/* ... */);
    // instance.owner may be invalid now, but you can still:
    instance.map_aliased(/* ... */);
    instance.map_aliased(/* ... */); // ...for multiple times
}
```

Maybe this needs to be worded better, now that the `once` method is gone? But I'm not sure if I know a better wording.

Reply:
Oh okay, in the other doc-comments it said "mutliple times" as well, I assumed it was a typo.
Maybe an example like this can be added to the doc-comment?

Reply:
Added.